### PR TITLE
FXC-4967 fix(geometry): use relative tolerance in subdivide sliver filter

### PR DESCRIPTION
**Pull request by dmarek-flex**
Addresses FXC-4967: https://flow360.atlassian.net/browse/FXC-4967

The `subdivide` function was using an absolute area tolerance (`fp_eps ≈ 1.2e-7`) to filter out sliver polygons. For large geometries with coordinates in the thousands, this absolute tolerance was too lenient and allowed degenerate slivers to pass through. Changed to use a relative tolerance based on the original geometry's area.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> Prevents degenerate sliver polygons during 2D material subdivision by making filtering relative to grid cell size.
> 
> - Added `_is_sliver_polygon()` with grid-relative area/dimension thresholds (`_SLIVER_AREA_RTOL`, `_SLIVER_DIM_RTOL`) and updated `subdivide()` to accept optional `grid`
> - Updated `Simulation` to pass `grid` to `subdivide()` when expanding 2D materials
> - Added tests for large-coordinate sliver cases and grid-based filtering; updated changelog entry
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 22e6c49ed84cecdffee9fc6909abeb31d7fa89ef. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

Changed `subdivide()` function from using absolute area tolerance to relative tolerance based on grid cell size for filtering sliver polygons.

- Added `_SLIVER_AREA_RTOL` and `_SLIVER_DIM_RTOL` constants (both 1e-4) for relative tolerance thresholds
- Implemented `_is_sliver_polygon()` helper function that checks both area and bounding box dimensions relative to grid cell sizes
- Updated `subdivide()` to accept optional `grid` parameter for computing relative thresholds
- Modified simulation.py to pass grid to `subdivide()` when processing 2D materials
- Added comprehensive test case `test_subdivide_large_geometry_sliver_filter()` validating the fix with large coordinates

The fix addresses floating-point precision issues with large geometries (coordinates ~7000) where the previous absolute tolerance (`fp_eps` ≈ 1.2e-7) was insufficient. With relative tolerance, slivers smaller than 0.01% of grid cell size are correctly filtered out.

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with minimal risk
- The implementation is well-designed with proper backward compatibility (optional grid parameter), comprehensive test coverage for the specific bug scenario, clear documentation, and follows the codebase conventions. The logic correctly handles edge cases (None grid, zero dimensions) and the relative tolerance approach is sound for addressing floating-point precision issues at large coordinates.
- No files require special attention

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| tidy3d/components/geometry/utils_2d.py | Added `_is_sliver_polygon()` helper function and updated `subdivide()` to accept optional grid parameter for relative tolerance-based sliver filtering |
| tidy3d/components/simulation.py | Updated `subdivide()` call to pass grid parameter for grid-aware sliver filtering |
| tests/test_components/test_geometry.py | Added comprehensive test for grid-based sliver filtering with large geometry coordinates |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Sim as Simulation
    participant Sub as subdivide()
    participant Filter as _is_sliver_polygon()
    participant Grid as Grid

    Sim->>Sub: subdivide(geometry, structures, grid)
    Note over Sub: Perform shapely boolean operations<br/>on 2D geometry intersections
    
    loop For each resulting polygon
        Sub->>Filter: _is_sliver_polygon(polygon, axis, grid)
        
        Filter->>Filter: Check area < _MIN_POLYGON_AREA
        alt Area below absolute minimum
            Filter-->>Sub: True (filter out)
        else Grid is None
            Filter-->>Sub: False (keep)
        else Grid provided
            Filter->>Grid: Get min cell sizes in tangential axes
            Filter->>Filter: Calculate min_cell_area
            Filter->>Filter: Check area < min_cell_area * _SLIVER_AREA_RTOL
            alt Area too small relative to grid
                Filter-->>Sub: True (filter out)
            else
                Filter->>Filter: Check bbox dims < min_cell_size * _SLIVER_DIM_RTOL
                alt Any dimension too small
                    Filter-->>Sub: True (filter out)
                else
                    Filter-->>Sub: False (keep)
                end
            end
        end
    end
    
    Sub-->>Sim: List of filtered subdivisions
```

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 - **Do** reuse `scripts/` utilities; **don't** add new helper modules without checking for an existing script first.
 
 ## Coding Style & Naming
+- Place all imports at the top of files, not inside functions; use absolute imports, not relative imports.
```
Comment (weiliangjin2021) [inline AGENTS.md]: does this forbid lazy import?

**Review by weiliangjin2021 — COMMENTED** [review]
Looks nice. I'll test it out now

File: AGENTS.md (hunk 1 above)
Comment (dmarek-flex) [reply AGENTS.md]: It doesn't forbid it, more like a default. Likely will remove

Comment (dmarek-flex) [reply AGENTS.md]: fixed in the lumped element snapping PR

**Review by weiliangjin2021 — APPROVED** [review]
